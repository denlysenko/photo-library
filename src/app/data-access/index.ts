export * from './photos/photos.service';
