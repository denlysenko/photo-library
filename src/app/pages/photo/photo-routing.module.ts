import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';

import { PhotoComponent } from './photo.component';
import { PhotoResolver } from './photo.resolver';

@NgModule({
  imports: [
    RouterModule.forChild([
      {
        path: '',
        component: PhotoComponent,
        resolve: {
          photo: PhotoResolver,
        },
      },
    ]),
  ],
  exports: [RouterModule],
})
export class PhotoRoutingModule {}
