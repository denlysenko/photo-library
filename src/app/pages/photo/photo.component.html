<div *ngIf="photo$ | async as photo"
     class="photo">
  <app-photo-tile [photo]="photo"
                  size="lg"></app-photo-tile>
  <button type="button"
          class="btn btn-warn"
          [disabled]="loading$ | async"
          (click)="removeFromFavorites(photo.id)"
          data-testid="remove-button">Remove From Favorites</button>
</div>