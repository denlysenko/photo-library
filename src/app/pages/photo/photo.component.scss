:host {
  width: 100%;
}

.photo {
  display: flex;
  flex-direction: column;
  align-items: center;
}
