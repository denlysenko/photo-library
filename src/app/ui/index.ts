export * from './header/header.module';
export * from './photo-tile/photo-tile.module';
export * from './photo-grid/photo-grid.module';
export * from './fullpage-spinner/fullpage-spinner.module';
export * from './infinite-scroll/infinite-scroll.module';
export * from './inline-spinner/inline-spinner.module';
