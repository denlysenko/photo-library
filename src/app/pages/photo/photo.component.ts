import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';

import { Observable } from 'rxjs';
import { finalize, map } from 'rxjs/operators';

import { ProgressComponent } from '../../base';
import { PhotosService } from '../../data-access';
import { Photo } from '../../interfaces';

@Component({
  selector: 'app-photo',
  templateUrl: './photo.component.html',
  styleUrls: ['./photo.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class PhotoComponent extends ProgressComponent {
  photo$: Observable<Photo> = this.route.data.pipe(map((data) => data.photo));

  constructor(
    private readonly route: ActivatedRoute,
    private readonly photosService: PhotosService,
    private readonly router: Router
  ) {
    super();
  }

  removeFromFavorites(photoId: string): void {
    this.loading.next(true);

    this.photosService
      .removeFromFavorites(photoId)
      .pipe(finalize(() => this.loading.next(false)))
      .subscribe(() => {
        this.router.navigate(['/favorites']);
      });
  }
}
