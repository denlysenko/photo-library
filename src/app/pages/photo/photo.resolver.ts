import { Injectable } from '@angular/core';
import { ActivatedRouteSnapshot, Resolve } from '@angular/router';

import { Observable } from 'rxjs';

import { PhotosService } from '../../data-access';
import { Photo } from '../../interfaces';

@Injectable()
export class PhotoResolver implements Resolve<Photo> {
  constructor(private readonly photosService: PhotosService) {}

  resolve(route: ActivatedRouteSnapshot): Observable<Photo> {
    const photoId = route.paramMap.get('id');
    return this.photosService.getFavoriteById(photoId);
  }
}
