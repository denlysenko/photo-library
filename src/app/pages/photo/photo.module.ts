import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { PhotoTileModule } from '../../ui';
import { PhotoRoutingModule } from './photo-routing.module';
import { PhotoComponent } from './photo.component';
import { PhotoResolver } from './photo.resolver';

@NgModule({
  imports: [CommonModule, PhotoRoutingModule, PhotoTileModule],
  declarations: [PhotoComponent],
  providers: [PhotoResolver],
})
export class PhotoModule {}
