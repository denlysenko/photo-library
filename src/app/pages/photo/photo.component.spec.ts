import { CommonModule } from '@angular/common';
import {
  async,
  ComponentFixture,
  TestBed,
  fakeAsync,
  tick,
} from '@angular/core/testing';
import { ActivatedRoute, Router } from '@angular/router';

import { of } from 'rxjs';

import { PhotosService } from '../../data-access';
import { PhotoTileModule } from '../../ui';
import { PhotoComponent } from './photo.component';

const photo = {
  id: 'uuid',
  link: 'link',
};

describe('PhotoComponent', () => {
  let component: PhotoComponent;
  let fixture: ComponentFixture<PhotoComponent>;
  let photosService: PhotosService;
  let router: Router;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [CommonModule, PhotoTileModule],
      declarations: [PhotoComponent],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: {
            data: of({
              photo,
            }),
          },
        },
        {
          provide: Router,
          useValue: {
            navigate: jest.fn(),
          },
        },
        {
          provide: PhotosService,
          useValue: {
            removeFromFavorites: jest.fn(() => of(true)),
          },
        },
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(PhotoComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    photosService = TestBed.inject(PhotosService);
    router = TestBed.inject(Router);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('removeFromFavorites', () => {
    it('should remove photo from favorites', fakeAsync(() => {
      fixture.debugElement.nativeElement
        .querySelector('[data-testid=remove-button]')
        .click();

      tick();

      expect(photosService.removeFromFavorites).toHaveBeenCalledWith(photo.id);
    }));

    it('should navigate to favorites page', fakeAsync(() => {
      fixture.debugElement.nativeElement
        .querySelector('[data-testid=remove-button]')
        .click();

      tick();

      expect(router.navigate).toHaveBeenCalledWith(['/favorites']);
    }));
  });
});
