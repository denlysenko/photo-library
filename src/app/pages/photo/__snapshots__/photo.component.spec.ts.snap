// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoComponent should create 1`] = `
<app-photo
  loading={[Function BehaviorSubject]}
  loading$={[Function Observable]}
  photo$={[Function Observable]}
  photosService={[Function Object]}
  route={[Function Object]}
  router={[Function Object]}
>
  <div
    class="photo"
  >
    <app-photo-tile
      size="lg"
    >
      <div
        class="photo-tile card lg"
      >
        <img
          alt="photo"
          class="photo"
          src="link"
        />
      </div>
    </app-photo-tile>
    <button
      class="btn btn-warn"
      data-testid="remove-button"
      type="button"
    >
      Remove From Favorites
    </button>
  </div>
</app-photo>
`;
